engine/lib/ast_utils.ml

```diff
@@ -410,8 +410,12 @@ module Make (F : Features.T) = struct
           inherit [_] Visitors.reduce as super
           inherit [_] Sets.Local_ident.monoid as _m
 
-          method! visit_arm' env { arm_pat; body } =
-            shadows ~env [ arm_pat ] body super#visit_expr
+          method! visit_arm' env { arm_pat; body; guard } =
+            match guard with
```

Comment:
Here, you can remove duplicate code with a `shadows ~env [ arm_pat ] rhs super#visit_expr ++ (Option.map ~f:(...) guard |> Option.value ~default_or_thunk:self#zero))`

Reply:
I think it is not possible here as in the `None` branch the expression on which we call `shadows` is `body`. But maybe another simplification is possible

Reply:
Ah, right, that makes sense!